app/src/main/java/org/mozilla/focus/widget/FirefoxProgressBar.kt

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.widget
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.arch.lifecycle.Observer
+import android.os.Handler
+import android.os.Looper
+import android.os.Message
+import android.view.View
+import android.widget.LinearLayout
+import org.mozilla.focus.R
+import org.mozilla.focus.architecture.NonNullObserver
+import org.mozilla.focus.fragment.BrowserFragment
+import kotlinx.android.synthetic.main.firefox_progress_bar.view.*
+import java.lang.ref.WeakReference
+import java.util.concurrent.TimeUnit
+
+private const val HIDE_MESSAGE_ID = 0
+private const val HIDE_ANIMATION_DURATION_MILLIS = 250L
+private val HIDE_AFTER_MILLIS = TimeUnit.SECONDS.toMillis(3)
+
+class FirefoxProgressBar @JvmOverloads constructor(
+        context: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
+    : LinearLayout(context, attrs, defStyle) {
+
+    private val hideHandler = FirefoxProgressBarHideHandler(this)
+
+    fun initialize(browserFrag: BrowserFragment) {
+        browserFrag.session.url.observe(browserFrag, Observer { sessionUrl -> url.text = sessionUrl })
+        browserFrag.session.loading.observe(browserFrag, object : NonNullObserver<Boolean>() {
+            public override fun onValueChanged(loading: Boolean) {
+                if (loading) {
+                    startUpdates()
+                } else {
+                    resetCountdown()
+                }
+            }
+        })
+    }
+
+    init {
+        LayoutInflater.from(context)
+                .inflate(R.layout.firefox_progress_bar, this, true)
+        gifImageView.setGifImageResource(R.drawable.progress_loading_indicator)
+
+        // hardware acceleration needs to be disabled for gif support in FirefoxProgressBar
+        // http://android-er.blogspot.it/2014/03/play-animated-gif-with.html
+        setLayerType(View.LAYER_TYPE_SOFTWARE, null)
+    }
+
+    fun startUpdates() {
+        setMaxVisibility()
+        resetCountdown()
+    }
+
+    private fun setMaxVisibility() {
+        animate().cancel()
+        alpha = 1f
+    }
+
+    private fun resetCountdown() {
+        hideHandler.removeMessages(HIDE_MESSAGE_ID)
```

Comment:
nit: I'm not sure this hide handler is necessary – you might be able to call `animate().setStartDelay(...)` directly. The reason I do it for the cursor is because every time the user presses the directional buttons, we'll allocate and throw away some objects, which can be really bad depending on how much they tap it. In this case, I don't think we loading=false will get called often but I could be wrong (e.g. page loads so =false, then javascript runs, starts page loads again, etc.). I'm not sure how WebView will handle these cases.

It seems like the safer approach is to leave it as is though. :)